### ISSUE TITLE
Need the OnEnter() and OnExit() to be run even if when transitioning to and from the same state.

### ISSUE THREAD

**Issue by king0952** [bevyengine/bevy#9130]
## What problem does this solve or what need does it fill?

When using bevy 0.10.1, I can trigger a button click event and enter StateA, and query some resources and spawn some bundles when I enter this StateA with `add_system(setup.in_schedule(OnEnter(StateA)))`, every time I click the button (without leave the StateA), I can despawn bundles by using `add_system(exit.in_schedule(OnExit(StateA)));` and  spawn new bundles according to the resources again. But after I update to bevy 0.11, this cannot be done because of [#8359](https://github.com/bevyengine/bevy/pull/8359).

## What solution would you like?

I hope `OnEnter` and `OnExit` can be run again when the system enters the same state, or at least, can make it optional.

## What alternative(s) have you considered?

No effective methods were found.

**Comment by Luminoth** [comment]
Re-entering states is definitely a useful thing to be able to do, having this functionality back in some form would be really great to see.

**Comment by vacuus** [comment]
Add `OnEachEnter` and `OnEachExit` schedules?

**Comment by Selene-Amanita** [comment]
> Add `OnEachEnter` and `OnEachExit` schedules?

I think some people might not expect their state to be reentered andit would create some bugs if simply re-added to `OnEnter`. Having an alternative solution like this is better.

**Comment by rlidwka** [comment]
`OnEnter` should be triggered when re-entering the same state.

If a user doesn't want that to happen, she needs a way to check the previous state (not sure if previous state is kept track of now).

**Comment by vacuus** [comment]
There are use cases for both re-triggering and not re-triggering. See https://github.com/bevyengine/bevy/issues/8191. I don't know what the overhead of a schedule is in Bevy, but if it's not significant then both use cases should be easily supported. The docs should be clear on the distinction. I think it's not really cognitive overhead; the state transition logic maps quite nicely to the schedules.

**Comment by nightactive-git** [comment]
I would also highly appreciate if the state transition is triggered, also if the next state is the same as the current state. State transitions are a very powerful and robust mechanism to execute a bunch of actions/systems in a specific order (or in parallel).
Workarounds are painful and cause a lot of clumsy code, which (in my option) doesn't harmonize with the spirit of an ECS engine.
I would really appreciate a hot fix for that.

**Comment by urben1680** [comment]
An associated constant `bool` on the `States` trait (and it's derive) could be the simple fix here that determines what should happen if the `StateNext<S>` resource is updated with a new state that is equal to the inner value of the resource `State<S>`.

A hacky workaround until a solution for Bevy has been deployed can be to manually implement `PartialEq` for your state type and make `eq` always return false.

**Comment by nightactive-git** [comment]
Thank you @urben1680 for this advice.
When I override the function eq so that it always returns false, the state transition doesn't seem to work anymore. Is there anything else, I have to regard?

`impl PartialEq for ToolbarState {
    fn eq(&self, &other: &Self) -> bool {
        return false;
    }
}`

**Comment by urben1680** [comment]
@nightactive-git Oh too bad, I did not test it as it seemed obvious to me that it would work. Sorry.

**Comment by urben1680** [comment]
@nightactive-git This might work better, I only tested it if it compiles:

```rs
#[derive(Clone, PartialEq, Eq, Hash, Debug, Default, States)]
pub enum MyStateInner{
    #[default]
    StateA,
    StateB
}

#[derive(Clone, PartialEq, Eq, Hash, Debug, Default)]
pub struct MyState{
    state: MyStateInner,
    flip: bool
}

/// derive only supports fieldless enums so let's do this
impl States for MyState{
    type Iter = std::array::IntoIter<Self, 4>;
    fn variants() -> Self::Iter {
        [
            Self { state: MyStateInner::StateA, flip: false },
            Self { state: MyStateInner::StateA, flip: true },
            Self { state: MyStateInner::StateB, flip: false },
            Self { state: MyStateInner::StateB, flip: true }
        ].into_iter()
    }
}

impl MyState {
    /// use this to construct a state using the current state
    pub fn unequal_new(&self, state: MyStateInner) -> Self {
        Self {
            state,
            flip: !self.flip
        }
    }
}
```

**Comment by alice-i-cecile** [comment]
I'm in favor of supporting this as an opt-in behavior: it's clearly useful, but still quite surprising.

**Comment by benfrankel** [comment]
Made a PR re-adding the behavior as opt-in via `NextState::set_forced`.

**Comment by MiniaczQ** [comment]
I don't think there is a need to complicate `NextState` with an opt-in API, I'd much better prefer a `SystemParam` wrapper around `(State, NextState)` pair that performs the check for me. It's less invasive.

**Comment by benfrankel** [comment]
> I don't think there is a need to complicate `NextState` with an opt-in API, I'd much better prefer a `SystemParam` wrapper around `(State, NextState)` pair that performs the check for me. It's less invasive.

That would imply that using `NextState` directly would have the reverted behavior, and you'd have to use the `SystemParam` instead for the current behavior. (this isn't an argument for/against just clarifying)

Also, if you e.g. call `param.set(foo)` followed by `param.set(bar)` in the same system, while the current state is `bar`, this would queue a transition to `foo`.

**Comment by MiniaczQ** [comment]
The wrapper can check for values inside of `NextState` as well, the fun part is that it can be as complicated as you want because it doesn't mess up the simpler state logic

**Comment by benfrankel** [comment]
Okay I see what you're saying. You could give `param.set` the current behavior by setting `next_state.0 = None` instead of `Some(current_state)`, and `param.set_forced` can have the old behavior by setting `Some(current_state)`.

**Comment by MiniaczQ** [comment]
Few more thoughts:

The `NextState` force flag idea won't work with cases like:
```rs
enum FooBar {
  #[default]
  Foo,
  Bar,
}

app.add_systems(Update, (sys_a, sys_b).chain());

fn sys_a(next: ResMut<NextState<FooBar>>) {
  // Should set Foo to Bar
  nest.set_force(FooBar::Bar);
}

fn sys_b(next: ResMut<NextState<FooBar>>) {
  // Replaces Bar, no state change occurs
  nest.set_force(FooBar::Foo);
}
```
which *could* be the intended way for this, but definitely not the one I'd expect.

**Comment by MiniaczQ** [comment]
Another problem with the entire idea of forced updates is it straight up doesn't make sense when you nest states.
Which state is meant to be 'forced'? You can't decide that during `NextState::set` like in the suggestions, you can only set it during compilation by replacing `Eq` implementation, which  may, and most likely will mess up any derived implementations.

I think we shouldn't support forced updates, at least not like this.

**Comment by benfrankel** [comment]
> Few more thoughts:
> 
> The `NextState` force flag idea won't work with cases like:
> 
> ```rust
> ...
> ```
> 
> which _could_ be the intended way for this, but definitely not the one I'd expect.

It is the intended behavior. If it's unexpected, that's probably because of the name of the method. It's supposed to convey that the state transition will happen (if it's still queued when `apply_state_transitions` runs) regardless of the `from` state -- but you could also read it as conveying that the state transition will happen regardless of other calls to `next_state.set`.

**Comment by MiniaczQ** [comment]
If the current state is `Foo`, then I'd expect a `set` to be ignored if I try to use `Foo`, but proceed with an update if it's anything else, like `Bar`.
The problem here is that the state change is deferred, so your decision that would succeed is overwritten by a decision that will fail. Hence resolving it immediately seems more reasonable to me.

**Comment by pablo-lua** [comment]
The Idea of removing the verification and let the user verifying themselves if they are trying to change something to the same state would probably be the simplest and a good way to do this?

**Comment by benfrankel** [comment]
I don't think setting the state to `Foo` should be fully ignored if the current state is `Foo`. There was a logical intention for the next state to be `Foo`, whether that means nothing should happen because the state is already `Foo`, or there should be an `OnEnter(Foo)` / `OnExit(Foo)`, either way, you would expect the state to be `Foo` in the next frame.

**Comment by benfrankel** [comment]
> The Idea of removing the verification and let the user verifying themselves if they are trying to change something to the same state would probably be the simplest and a good way to do this?

Yeah that effectively lines up with @MiniaczQ's proposal which I agree with. You'd choose between `ChangeState::set` and `ChangeState::set_forced` where `ChangeState { current: Res<State<S>>, next: ResMut<NextState<S>> }`.

**Comment by MiniaczQ** [comment]
> There was a logical intention for the next state to be `Foo`

What if this happened due to system order ambiguity, then it's an additional thing to worry about

**Comment by benfrankel** [comment]
> What if this happened due to system order ambiguity, then it's an additional thing to worry about

If you have a race condition between systems that set the next state, that's just a logic error. I don't think there's a consistently sane way for bevy to handle that.

**Comment by MiniaczQ** [comment]
Just resolving whether a state update should occur during `set`, instead of the `transition` system would fix this

**Comment by benfrankel** [comment]
Made a PR for the system param approach, which I prefer. I'm leaving the `NextState::set_forced` PR open for now.

**Comment by benfrankel** [comment]
Until this is fixed, the easiest workaround is to add a `RestartFoo` state and a system upon entering `RestartFoo` to set the next state to `Foo`. This introduces a 1 frame delay and a little bit of boilerplate, but it should be easy to migrate to actually re-entering `Foo` when that becomes possible.

**Comment by benfrankel** [comment]
This is fixed in my 3rd-party crate [pyri_state](https://github.com/benfrankel/pyri_state). I was unable to get any fix upstreamed unfortunately because there's more than one way to implement it and there are disagreements.
